File: Dynamic Programming/377. Combination Sum IV.cpp
/*
backtracking 会time out
*/

/*
DP
*/

class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        vector<int>dp(target+1,0);
        dp[0] = 1;
        sort(nums.begin(),nums.end());
        for(int i = 1; i<=target;i++){
            for(int j = 0; j<nums.size() && nums[j]<=i; j++)
                dp[i] += dp[i-nums[j]];
        }
        return dp[target];
    }
};



//2020 Top-Down
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        vector<long long>dp(target+1,-1);
        dp[0] = 1;
        return topDown(nums, dp, target);
    }
    
    int topDown(vector<int>&nums, vector<long long>&dp, int target){
        if(dp[target] != -1)
            return dp[target];
        int cur = 0;
        for(auto j: nums)
            if(j<=target)
                cur += topDown(nums,dp, target-j);
        return dp[target] = cur;
    }
};

//2020 Bottom-Up
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        vector<long long>dp(target+1,0);
        dp[0] = 1;
        for(int i = 1; i<=target; ++i){
            int cur = 0;
            for(auto j: nums){
                if(j <= i && cur + dp[i-j]<numeric_limits<int>::max())//个别数很大，阻止overflow
                    cur += dp[i-j];
            }
            dp[i] = cur;
        }
        return dp.back();
    }
};

File: Dynamic Programming/377. Combination Sum IV.py
"""
377. Combination Sum IV
"""

class Solution:
    def combinationSum4(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        dp = [0 for _ in range(target+1)]
        for i in range(1,target+1):
            for n in nums:
                if n>i:
                    break
                if n == i:
                    dp[i] += 1
                else:
                    dp[i] += dp[i-n]
        return dp[target]

#bottom-up
class Solution:
    def combinationSum4(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        dp = [0]*(target + 1)
        dp[0] = 1
        for i in range(1, target+1):
            for num in nums:
                if num>i:
                    break
                dp[i]+=dp[i-num]
        return dp[target]
    

#bottom-up
class Solution:
    def combinationSum4(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        cache = {}
        def f(val):
          print(" in ",val)
          if val == target:
            return 1

          total = 0
          remain = target - val
          for num in nums:
            if num <= remain:
              k = val+num
              if k in cache:
                total += cache[k]
              else:
                cache[k] = f(val + num)
                total += cache[k]
          return total
        
        return f(0)

#2020 Top-Down
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = [1] + [-1 for _ in range(target)]
        def topDown(target):
            if dp[target] >= 0:
                return dp[target]
            cur = 0
            for i in nums:
                if i <= target:
                    cur += topDown(target-i)
            dp[target] = cur
            return dp[target]
        return topDown(target)